carddav/server.go

```diff
@@ -25,8 +25,18 @@ type PutAddressObjectOptions struct {
 	IfMatch string
 }
 
+type ResourceType int
+
+const (
+	ResourceTypeUserPrincipal      ResourceType = 1 << iota
+	ResourceTypeAddressBookHomeSet ResourceType = 1 << iota
+	ResourceTypeAddressBook        ResourceType = 1 << iota
+	ResourceTypeAddressObject      ResourceType = 1 << iota
+)
+
 // Backend is a CardDAV server backend.
 type Backend interface {
+	CardDAVResourceType(ctx context.Context, path string) (ResourceType, error)
```

Comment:
One thing I was unsure about: I would have just named this `ResourceType()`, but we'd need the same thing for CalDAV eventually, and it's better to allow one type to implement both interfaces (i.e. keep the interfaces conflict-free).

One option might be to make this generic enough, e.g.: user principal, home set, collection, object - and then use that for both? Not sure, let me know if you have ideas...